pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go

```diff
@@ -1841,242 +1842,312 @@ func getNumberOfPodInterfaces(spec *v1.VirtualMachineInstanceSpec) int {
 	return nPodInterfaces
 }
 
-func validateDisks(field *k8sfield.Path, disks []v1.Disk) []metav1.StatusCause {
+func validateDiskName(field *k8sfield.Path, idx int, disks []v1.Disk) []metav1.StatusCause {
 	var causes []metav1.StatusCause
-	nameMap := make(map[string]int)
-
-	for idx, disk := range disks {
-		// verify name is unique
-		otherIdx, ok := nameMap[disk.Name]
-		if !ok {
-			nameMap[disk.Name] = idx
-		} else {
+	for otherIdx, disk := range disks {
+		if otherIdx < idx && disk.Name == disks[idx].Name {
 			causes = append(causes, metav1.StatusCause{
 				Type:    metav1.CauseTypeFieldValueInvalid,
 				Message: fmt.Sprintf("%s and %s must not have the same Name.", field.Index(idx).String(), field.Index(otherIdx).String()),
 				Field:   field.Index(idx).Child("name").String(),
 			})
 		}
+	}
+	return causes
+}
 
-		// Verify only a single device type is set.
-		deviceTargetSetCount := 0
-		var diskType string
-		var bus v1.DiskBus
-		if disk.Disk != nil {
-			deviceTargetSetCount++
-			diskType = "disk"
-			bus = disk.Disk.Bus
-		}
-		if disk.LUN != nil {
-			deviceTargetSetCount++
-			diskType = "lun"
-			bus = disk.LUN.Bus
-		}
-		if disk.CDRom != nil {
+func validateDeviceTarget(field *k8sfield.Path, idx int, disk v1.Disk) []metav1.StatusCause {
+	var causes []metav1.StatusCause
+	deviceTargetSetCount := 0
+	for _, target := range []interface{}{disk.Disk, disk.LUN, disk.CDRom} {
+		if !reflect.ValueOf(target).IsNil() {
 			deviceTargetSetCount++
-			diskType = "cdrom"
-			bus = disk.CDRom.Bus
 		}
+	}
+	// NOTE: not setting a device target is okay. We default to Disk.
+	// However, only a single device target is allowed to be set at a time.
+	if deviceTargetSetCount > 1 {
+		causes = append(causes, metav1.StatusCause{
+			Type:    metav1.CauseTypeFieldValueInvalid,
+			Message: fmt.Sprintf("%s can only have a single target type defined", field.Index(idx).String()),
+			Field:   field.Index(idx).String(),
+		})
+	}
+	return causes
+}
 
-		// NOTE: not setting a device target is okay. We default to Disk.
-		// However, only a single device target is allowed to be set at a time.
-		if deviceTargetSetCount > 1 {
+func validatePciAddress(field *k8sfield.Path, idx int, disk v1.Disk) []metav1.StatusCause {
+	var causes []metav1.StatusCause
+	if disk.Disk != nil && disk.Disk.PciAddress != "" {
```

Comment:
You could change this to an early return to avoid nesting the if conditions below.